qiskit_nature/operators/second_quantization/spin_op.py

```diff
@@ -12,117 +12,479 @@
+        if self.spin != other.spin:
+            raise TypeError(f"Addition between spin {self.spin} and spin {other.spin} is invalid.")
 
-    def reduce(self, atol: Optional[float] = None, rtol: Optional[float] = None):
+        return SpinOp(
+            (
+                np.hstack((self._spin_array, other._spin_array)),
+                np.hstack((self._coeffs, other._coeffs)),
+            ),
+            spin=self.spin,
+        )
+
+    def compose(self, other):
+        # TODO: implement
         raise NotImplementedError
+
+    def mul(self, other: complex) -> "SpinOp":
+        if not isinstance(other, (int, float, complex)):
+            raise TypeError(
+                "Unsupported operand type(s) for *: 'SpinOp' and " f"'{type(other).__name__}'"
+            )
+
+        return SpinOp((self._spin_array, self._coeffs * other), spin=self.spin)
+
+    def adjoint(self) -> "SpinOp":
+        # Note: X, Y, Z are hermitian, therefore the dagger operation on a SpinOperator amounts
+        # to simply complex conjugating the coefficient.
+        return SpinOp((self._spin_array, self._coeffs.conjugate()), spin=self.spin)
+
+    def reduce(self, atol: Optional[float] = None, rtol: Optional[float] = None) -> "SpinOp":
+        if atol is None:
+            atol = self.atol
+        if rtol is None:
+            rtol = self.rtol
+
+        flatten_array, indexes = np.unique(
+            np.column_stack(self._spin_array), return_inverse=True, axis=0
+        )
+        coeff_list = np.zeros(len(self._coeffs), dtype=np.complex128)
+        for i, val in zip(indexes, self._coeffs):
+            coeff_list[i] += val
+        non_zero = [
+            i for i, v in enumerate(coeff_list) if not np.isclose(v, 0, atol=atol, rtol=rtol)
+        ]
+        if not non_zero:
+            return SpinOp(
+                (
+                    np.zeros((3, 1, self.register_length), dtype=np.int8),
+                    np.array([0], dtype=np.complex128),
+                ),
+                spin=self.spin,
+            )
+        new_array = flatten_array[non_zero].T.reshape((3, len(non_zero), self._register_length))
+        new_coeff = coeff_list[non_zero]
+        return SpinOp((new_array, new_coeff), spin=self.spin)
+
+    def __len__(self) -> int:
+        return len(self._coeffs)
+
+    def to_list(self) -> List[Tuple[str, complex]]:
+        """Getter for the list which represents `self`
+
+        Returns:
+            The list [(label, coeff)]
+        """
+        coeff_list = self._coeffs.tolist()
+        return [(self._generate_label(i), coeff_list[i]) for i in range(len(self))]
+
+    def _generate_label(self, i):
+        """Generates the string description of `self`."""
+        labels_list = []
+        for pos, (n_x, n_y, n_z) in enumerate(self._spin_array[:, i].T):
+            rev_pos = self.register_length - pos - 1
+            if n_x > 1:
+                labels_list.append(f"X_{rev_pos}^{n_x}")
+            if n_x == 1:
```

Comment:
Seems like we might use elif.

Reply:
Why?

Reply:
Currently, when n_x = 2 for example, Python will still evaluate 2 conditions to see if they are true or not. If the second one was elif, Python would only evaluate the first one.

Reply:
You are right. But , it doesn't have a big impact, and the ternary operator (conditional operator) is faster.
<img width="840" alt="image" src="https://user-images.githubusercontent.com/6814928/110266489-3eb05c00-8001-11eb-860b-fd9b7e15aad2.png">
<img width="805" alt="image" src="https://user-images.githubusercontent.com/6814928/110266496-440da680-8001-11eb-9c66-34e107fcb674.png">

Reply:
Ok, great job with time tests!